### PR TITLE
Add function2 4.1.0

### PR DESCRIPTION
**Pull request by IceflowRE**
Related #2070

Specify library name and version:  **function2/4.1.0**
Required by: #2076 

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/function2/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    license = "Boost Software License 1.0"
+    settings = "os", "compiler", "arch", "build_type"
+    no_copy_source = True
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def configure(self):
+        tools.check_min_cppstd(self, "14")
```
Comment (madebr) [inline recipes/function2/all/conanfile.py:21]:
```suggestion
        if self.settings.compiler.cppstd:
            tools.check_min_cppstd(self, 14)
```

File: recipes/function2/all/test_package/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,10 @@
+cmake_minimum_required(VERSION 3.11)
+project(test_package)
+
+set(CMAKE_CXX_STANDARD 14)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
```
Comment (madebr) [inline recipes/function2/all/test_package/CMakeLists.txt:6]:
```suggestion
project(test_package)

include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
```

File: recipes/function2/all/test_package/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+conan_basic_setup()
+
+add_executable(${PROJECT_NAME} test_package.cpp)
+target_link_libraries(${PROJECT_NAME} ${CONAN_LIBS})
```
Comment (madebr) [inline recipes/function2/all/test_package/CMakeLists.txt:10]:
```suggestion
target_link_libraries(${PROJECT_NAME} ${CONAN_LIBS})
set_property(TARGET ${PROJECT_NAME} PROPERTY CXX_STANDARD 14)
```

File: recipes/function2/all/conanfile.py (hunk 1 above)
Comment (IceflowRE) [reply recipes/function2/all/conanfile.py:21]: i have a question about it there are several options how it is done in center:
```python
if self.settings.get_safe("compiler.cppstd"):
    tools.check_min_cppstd(self, 14)
```
```python
if self.settings.compiler.get_safe("cppstd"):
    tools.check_min_cppstd(self, 14)
```
```python
if self.settings.compiler.cppstd:
    tools.check_min_cppstd(self, 14)
```
```python
tools.check_min_cppstd(self, 14)
```

Even if using a number or a string is different. What should now be used and why?

Comment (madebr) [reply recipes/function2/all/conanfile.py:21]: It does not matter, inside the function it is *stringized*.

File: recipes/function2/all/test_package/CMakeLists.txt (hunk 2 above)
Comment (IceflowRE) [reply recipes/function2/all/test_package/CMakeLists.txt:6]:
```cpp
set_target_properties(
    ${PROJECT_NAME} PROPERTIES

    CXX_STANDARD 14
    CXX_STANDARD_REQUIRED ON
)
```
Done.

File: recipes/function2/all/test_package/CMakeLists.txt (hunk 3 above)
Comment (IceflowRE) [reply recipes/function2/all/test_package/CMakeLists.txt:10]:
```cpp
set_target_properties(
    ${PROJECT_NAME} PROPERTIES

    CXX_STANDARD 14
    CXX_STANDARD_REQUIRED ON
)
```
Done.

**Comment by IceflowRE** [comment]
Removed settings attribute.

**Comment by IceflowRE** [comment]
Simplified settings attribute.

File: recipes/function2/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+from conans import ConanFile, tools
+
+
+class Function2Conan(ConanFile):
+    name = "function2"
+    description = "Improved and configurable drop-in replacement to std::function that supports move only types, multiple overloads and more"
+    topics = "function", "header-only"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/Naios/continuable"
+    license = "Boost Software License 1.0"
```
Comment (uilianries) [inline recipes/function2/all/conanfile.py:12]:
```suggestion
    license = "BSL-1.0"
```

File: recipes/function2/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+from conans import ConanFile, tools
+
+
+class Function2Conan(ConanFile):
+    name = "function2"
+    description = "Improved and configurable drop-in replacement to std::function that supports move only types, multiple overloads and more"
+    topics = "function", "header-only"
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/Naios/continuable"
```
Comment (uilianries) [inline recipes/function2/all/conanfile.py:11]:
```suggestion
    homepage = "https://github.com/Naios/function2"
```

File: recipes/function2/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
+import os
+
+from conans import ConanFile, tools
+
+
+class Function2Conan(ConanFile):
+    name = "function2"
+    description = "Improved and configurable drop-in replacement to std::function that supports move only types, multiple overloads and more"
+    topics = "function", "header-only"
```
Comment (uilianries) [inline recipes/function2/all/conanfile.py:9]:
```suggestion
    topics = ("function", "functional", "function-wrapper", "type-erasure", "header-only")
```

File: recipes/function2/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        tools.check_min_cppstd(self, "14")
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = "function2-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def package(self):
+        self.copy(pattern="LICENSE.txt", dst="licenses", src=self._source_subfolder)
+        self.copy(pattern="*", dst="include/function2", src=self._source_subfolder + "/include/function2")
```
Comment (uilianries) [inline recipes/function2/all/conanfile.py:30]:
```suggestion
        self.copy(pattern="*.hpp", dst=os.path.join("include", "function2"), src=os.path.join(self._source_subfolder, "include", "function2"))
```

File: recipes/function2/all/test_package/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,16 @@
+cmake_minimum_required(VERSION 3.11)
```
Comment (uilianries) [inline recipes/function2/all/test_package/CMakeLists.txt:1]:
```suggestion
cmake_minimum_required(VERSION 3.0.2)
```

File: recipes/function2/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    license = "Boost Software License 1.0"
+    settings = "compiler"
+    no_copy_source = True
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def configure(self):
+        tools.check_min_cppstd(self, "14")
```
Comment (uilianries) [inline recipes/function2/all/conanfile.py:21]:
```suggestion
    def configure(self):
        minimal_cpp_standard = "14"
        if self.settings.compiler.cppstd:
            tools.check_min_cppstd(self, minimal_cpp_standard)
        minimal_version = {
            "gcc": "5",
            "clang": "3.4",
            "apple-clang": "10",
            "Visual Studio": "14"
        }
        compiler = str(self.settings.compiler)
        if compiler not in minimal_version:
            self.output.warn(
                "%s recipe lacks information about the %s compiler standard version support" % (self.name, compiler))
            self.output.warn(
                "%s requires a compiler that supports at least C++%s" % (self.name, minimal_cpp_standard))
            return
        version = tools.Version(self.settings.compiler.version)
        if version < minimal_version[compiler]:
            raise ConanInvalidConfiguration("%s requires a compiler that supports at least C++%s" % (self.name, minimal_cpp_standard))
```

**Comment by IceflowRE** [comment]
May i squash after the three approvals?

**Comment by danimtb** [comment]
> May i squash after the three approvals?

don't worry, we merge it with squash

**Pull request merged**